polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -3192,13 +3243,16 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
                    << " ST: " << *ST << "\n";
     }
     mlir::Type ET;
-    if (auto ST = PT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>()) {
+    if (auto ST = PT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>())
       ET = ST.getBody()[fnum];
-    } else {
-      ET = PT.getElementType()
-               .cast<mlir::LLVM::LLVMArrayType>()
-               .getElementType();
-    }
+    else if (auto AT =
+                 PT.getElementType().dyn_cast<mlir::LLVM::LLVMArrayType>())
+      ET = AT.getElementType();
+    else if (auto MT = PT.getElementType().dyn_cast<MemRefType>())
+      ET = MT.getElementType();
+    else
+      llvm_unreachable("not implemented");
```

Comment:
```suggestion
    
    llvm_unreachable("not implemented");
```

Reply:
They are not equivalent. The suggested code would assert even if  `PT.getElementType()` is `LLVMStructType`, `LLVMStructType`, or `LLVMArrayType`.